RavenDB-18159 Added the option to provide the shard context for a quey in the client so it will execute the query only on relevant shards. Refactoring of processors so the names of protected fields are upper cased and introducing common method for creation of ShardedQueryCommand to the base query processor class.

arekpalinski opened the pull request:
### Issue link
https://issues.hibernatingrhinos.com/issue/RavenDB-18159/Add-an-option-to-select-the-shards-for-query-in-the-client

### Additional description

- Introducing `IShardedQueryContextBuilder` class that will allow to provide the shard context query parameter so the orchestrator will run a query only on the relevant shards.
- Small refactoring of query processors

### Type of change

- New sharding feature

### How risky is the change?

- Low 

### Backward compatibility

- Not relevant

### Is it platform specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works
 - Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- It has been verified by manual testing

### Testing by RavenDB QA team

- No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- No

### UI work

- No UI work is needed

arekpalinski committed:
RavenDB-18159 Added the option to provide the shard context for a query in the client so it will execute the query only on relevant shards. Refactoring of processors so the names of protected fields are upper cased and introducing common method for creation of ShardedQueryCommand to the base query processor class.

ayende commented on src/Raven.Client/Documents/Session/AbstractDocumentQuery.Sharding.cs:
```diff
@@ -0,0 +1,24 @@
+
+        object shardContext;
+
+        if (builderImpl.DocumentIds.Count == 1)
+            shardContext = builderImpl.DocumentIds.First();
+        else
+            shardContext = builderImpl.DocumentIds;
```
Why do this? We can just pass an array in all cases.

arekpalinski replied:
We can do it. I just thought that since we have the distinction of the parameter type on the server side:

https://github.com/ravendb/ravendb/blob/dcf9e05775a18de4a56f912621f453a089d4da15/src/Raven.Server/Documents/Sharding/Queries/ShardedQueryProcessor.cs#L58-L61

we'll take advantage of that on the client API as well. At least we'll test that code properly.

ayende replied:
I'm fine either way

karmeli87 commented on src/Raven.Server/Documents/Sharding/Queries/AbstractShardedQueryProcessor.cs:
```diff
@@ -27,35 +28,76 @@ public abstract class AbstractShardedQueryProcessor<TCommand, TResult, TCombined
+            _filteredShardIndexes = new HashSet<int>();
+            switch (filter)
+            {
+                case LazyStringValue id:
```
can this be a `LazyCompressedStringValue` ?

arekpalinski replied:
Not very likely but potentially it can. I added it.

arekpalinski committed:
RavenDB-18159 Making sure that LazyCompressedStringValue is handled as well

arekpalinski committed:
RavenDB-18159 Review fixes:
- using IEnumerable instead of ICollection
- renamed IShardedQueryContextBuilder to IQueryShardedContext builder
- changing to internal accesibility of ShardContextParameterName const
